sharding/contracts/validator_manager.v.py

```diff
@@ -110,3 +141,88 @@ def sample(block_number: num, shard_id: num, sig_index: num) -> address:
                 return addr
 
     return zero_addr
+
+
+# Attempts to process a collation header, returns True on success, reverts on failure.
+def add_header(header: bytes <= 4096) -> bool:
+    values = RLPList(header, [num, num, bytes32, bytes32, bytes32, address, bytes32, bytes32, bytes])
+    shard_id = values[0]
+    expected_period_number = values[1]
+    period_start_prevhash = values[2]
+    parent_collation_hash = values[3]
+    tx_list_root = values[4]
+    collation_coinbase = values[5]
+    post_state_root = values[6]
+    receipt_root = values[7]
+    sig = values[8]
+
+    # Check if the header is valid
+    assert shard_id >= 0
+    assert block.number >= self.period_length
+    assert expected_period_number == floor(decimal(block.number / self.period_length))
+    assert period_start_prevhash == blockhash(expected_period_number * self.period_length - 1)
+
+    # Check if this header already exists
+    entire_header_hash = sha3(header)
+    assert not self.collation_headers[shard_id][entire_header_hash].exists
+    # Check whether the parent exists.
+    # if (parent_collation_hash == 0), i.e., is the genesis,
+    # then there is no need to check.
+    if parent_collation_hash != as_bytes32(0):
+        assert self.collation_headers[shard_id][parent_collation_hash].exists
+    # Check the signature with validation_code_addr
+    collator_valcode_addr = self.sample(shard_id)
+    sighash = extract32(raw_call(self.sighasher_addr, header, gas=200000, outsize=32), 0)
+    assert extract32(raw_call(collator_valcode_addr, concat(sighash, sig), gas=self.sig_gas_limit, outsize=32), 0) == as_bytes32(1)
+
+    # Add the header
+    _score = self.collation_headers[shard_id][parent_collation_hash].score + 1
+    self.collation_headers[shard_id][entire_header_hash] = {
+        parent_collation_hash: parent_collation_hash,
+        score: _score,
+        exists: True
```

Comment:
A thought that I am not confident of:
Since `collation_headers` may take a lot of storage space, how about replacing `exists` field with expression `self.collation_headers[shard_id][entire_header_hash].score > 0` when we have to make sure that it exists?
(Not sure the trade-off between how the cost of gas and storage space)